storage/tests/test_isolation.rs

```diff
@@ -122,3 +130,335 @@ fn g0_update_conflicts_fail() {
+    let key_1: &StorageKey<'_, 48> = &StorageKey::Reference(StorageKeyReference::new(KEYSPACE_ID, ByteReference::new(&KEY_1)));
+    let value_0 = ByteArray::inline([0], 1);
+
+    let storage_path = create_tmp_dir();
+    let storage = setup_storage(&storage_path);
+
+    let snapshot_setup = storage.open_snapshot_write();
+    snapshot_setup.put_val(key_1.to_owned(), ByteArray::copy(value_0.bytes()));
+    let result_snapshot = snapshot_setup.commit();
+    assert!(result_snapshot.is_ok());
+
+    let snapshot_1 = storage.open_snapshot_write();
+    let snapshot_2 = storage.open_snapshot_write();
+    let read_1: ByteArray<128> = snapshot_1.get::<128>(key_1).unwrap();
+    let read_2: ByteArray<128> = snapshot_1.get::<128>(key_1).unwrap();
+    let to_write_1 = ByteArray::inline([read_1.bytes()[0] + 1], 1);
+    let to_write_2 = ByteArray::inline([read_1.bytes()[0] + 1], 1);
+    // let to_write_1 = read_1.bytes()
+    // snapshot_1.put_val(key_1.to_owned(), );
+    // let read_12 = snapshot_2.get::<128>(key_1);
+    snapshot_1.put_val(key_1.to_owned(), ByteArray::copy(to_write_1.bytes()));
+    snapshot_2.put_val(key_1.to_owned(), ByteArray::copy(to_write_2.bytes()));
+    let result_1 = snapshot_1.commit();
+    let result_2 = snapshot_2.commit();
+
+    if (result_1.is_ok() && result_2.is_ok()) {
+        let snapshot_verify = storage.open_snapshot_read();
+        let read_verify: ByteArray<128> = snapshot_verify.get::<128>(key_1).unwrap();
+        assert_eq!(2, read_verify.bytes()[0]); // This does fail
+    }
+}
+
+#[test]
+fn g_single_read_skew() {
+    init_logging();
+    let key_1: &StorageKey<'_, 48> = &StorageKey::Reference(StorageKeyReference::new(KEYSPACE_ID, ByteReference::new(&KEY_1)));
+    let key_2: &StorageKey<'_, 48> = &StorageKey::Reference(StorageKeyReference::new(KEYSPACE_ID, ByteReference::new(&KEY_2)));
+    let value_10 = ByteArray::inline([10], 1); // value_<k><t> = Key k written by Transaction t
+    let value_12 = ByteArray::inline([12], 1);
+    let value_20 = ByteArray::inline([20], 1);
+    let value_22 = ByteArray::inline([22], 1);
+    let storage_path = create_tmp_dir();
+    let storage = setup_storage(&storage_path);
+
+    let snapshot_setup = storage.open_snapshot_write();
+    snapshot_setup.put_val(key_1.to_owned(), ByteArray::copy(value_10.bytes()));
+    snapshot_setup.put_val(key_2.to_owned(), ByteArray::copy(value_20.bytes()));
+    let result_snapshot = snapshot_setup.commit();
+    assert!(result_snapshot.is_ok());
+
+    let snapshot_1 = storage.open_snapshot_write();
+    let snapshot_2 = storage.open_snapshot_write();
+
+    let read_11 = snapshot_1.get::<128>(key_1);
+    snapshot_2.put_val(key_1.to_owned(), ByteArray::copy(value_12.bytes()));
+    snapshot_2.put_val(key_2.to_owned(), ByteArray::copy(value_22.bytes()));
+    let read_21 = snapshot_1.get::<128>(key_2);
+
+    assert_eq!(value_10.bytes(), read_11.unwrap().bytes());
+    assert_eq!(value_20.bytes(), read_21.unwrap().bytes());
```

Comment:
t1 reads key_1 before t2 writes, but reads key_2 after t2 writes. In our case, t1 actually reads both values before t2 writes anything.

(But then if t2 now commits, the reads would no longer be valid and t1 may have to be aborted)